test/functional/lua/fs_spec.lua

```diff
@@ -308,26 +310,133 @@ describe('vim.fs', function()
       )
     end)
 
-    it('works with UNC paths', function()
-      eq('//foo', vim.fs.normalize('//foo')) -- UNC path
-      eq('//foo/bar', vim.fs.normalize('//foo//bar////')) -- UNC path
-      eq('/foo', vim.fs.normalize('///foo')) -- Not a UNC path
-      eq('/', vim.fs.normalize('//')) -- Not a UNC path
-      eq('/', vim.fs.normalize('///')) -- Not a UNC path
-      eq('/foo/bar', vim.fs.normalize('/foo//bar////')) -- Not a UNC path
+    -- Opts required for testing posix paths and win paths
+    local posix_opts = is_os('win') and { win = false } or {}
+    local win_opts = is_os('win') and {} or { win = true }
+
+    it('preserves leading double slashes in POSIX paths', function()
+      eq('//foo', vim.fs.normalize('//foo', posix_opts))
+      eq('//foo/bar', vim.fs.normalize('//foo//bar////', posix_opts))
+      eq('/foo', vim.fs.normalize('///foo', posix_opts))
+      eq('//', vim.fs.normalize('//', posix_opts))
+      eq('/', vim.fs.normalize('///', posix_opts))
+      eq('/foo/bar', vim.fs.normalize('/foo//bar////', posix_opts))
     end)
 
-    if is_os('win') then
-      it('Last slash is not truncated from root drive', function()
-        eq('C:/', vim.fs.normalize('C:/'))
+    it('allows backslashes on unix-based os', function()
+      eq('/home/user/hello\\world', vim.fs.normalize('/home/user/hello\\world', posix_opts))
+    end)
+
+    it('preserves / after drive letters', function()
+      eq('C:/', vim.fs.normalize([[C:\]], win_opts))
+    end)
+
+    it('works with UNC and DOS device paths', function()
+      eq('//server/share/foo/bar', vim.fs.normalize([[\\server\share\foo\bar]], win_opts))
+      eq('//system07/C$/', vim.fs.normalize([[\\system07\C$\]], win_opts))
+      eq('//./C:/foo/bar', vim.fs.normalize([[\\.\C:\foo\bar]], win_opts))
+      eq('//?/C:/foo/bar', vim.fs.normalize([[\\?\C:\foo\bar]], win_opts))
+      eq(
+        '//?/UNC/server/share/foo/bar',
+        vim.fs.normalize([[\\?\UNC\server\share\foo\bar]], win_opts)
+      )
+      eq('//./BootPartition/foo/bar', vim.fs.normalize([[\\.\BootPartition\foo\bar]], win_opts))
+      eq(
+        '//./Volume{12345678-1234-1234-1234-1234567890AB}/foo/bar',
+        vim.fs.normalize([[\\.\Volume{12345678-1234-1234-1234-1234567890AB}\foo\bar]], win_opts)
+      )
+    end)
+
+    it('errors on invalid UNC and DOS device paths', function()
+      matches(
+        'Invalid Windows UNC path',
+        pcall_err(vim.fs.normalize, [[\\server\share]], win_opts)
+      )
+      matches(
+        'Invalid Windows UNC path',
+        pcall_err(vim.fs.normalize, [[\\server\]], win_opts)
+      )
```

Comment:
Not really sure we should be "validating" the path, for most of these cases. normalize() just normalizes what it's given. If it can produce a deterministic result, it should.